src/field/field.rs

```diff
@@ -486,3 +486,212 @@ macro_rules! test_square_root {
+            // Compare expected outputs with actual outputs
+            assert!(output.zip(expected).all(|(x, y)| x == y),
+                    "output differs from expected at rotation {}", i);
+        }
+        Ok(())
+    }
+}
+
+#[macro_export]
+macro_rules! test_arithmetic {
+    ($field:ty) => {
+        use crate::field_tests;
+        use crate::field_to_biguint;
+        use std::io::Result;
+        use std::ops::{Add,Sub,Mul,Neg,Div};
+        use num::{BigUint, Zero, One};
+
+
+        /// Return the modulus of the type `Fld`.
+        fn field_modulus<Fld>() -> BigUint
+        where Fld: Field {
+            field_to_biguint(Fld::T) * (BigUint::one() << Fld::TWO_ADICITY) + 1u32
+        }
+
+        // Can be 32 or 64; doesn't have to be computer's actual word
+        // bits. Choosing 32 gives more tests...
+        const WORD_BITS: usize = 32;
+
+        // TODO: I've given these unreasonably long names to avoid
+        // name conflicts (notably with the 'negation()' test for
+        // Bls12377Base. It would be better that these tests were
+        // simply in their own module.
+        #[test]
+        fn arithmetic_addition() -> Result<()> {
+            let modulus = field_modulus::<$field>();
+            field_tests::run_binaryop_test_cases(
+                &modulus, WORD_BITS,
+                <$field>::add,
+                |x, y| {
+                    let z = x + y;
+                    if z < modulus { z } else { z - &modulus }
+                })
+        }
+
+        #[test]
+        fn arithmetic_subtraction() -> Result<()> {
+            let modulus = field_modulus::<$field>();
+            field_tests::run_binaryop_test_cases(
+                &modulus, WORD_BITS,
+                <$field>::sub,
+                |x, y| if x >= y { x - y } else { &modulus - y + x })
+        }
+
+        #[test]
+        fn arithmetic_negation() -> Result<()> {
+            let modulus = field_modulus::<$field>();
+            field_tests::run_unaryop_test_cases(
+                &modulus, WORD_BITS,
+                <$field>::neg,
+                |x| if x.is_zero() { BigUint::zero() } else { &modulus - x })
+        }
+
+        #[test]
+        fn arithmetic_multiplication() -> Result<()> {
+            let modulus = field_modulus::<$field>();
+            field_tests::run_binaryop_test_cases(
+                &modulus, WORD_BITS,
+                <$field>::mul,
+                |x, y| x * y % &modulus)
+        }
+
+        #[test]
+        fn arithmetic_square() -> Result<()> {
+            let modulus = field_modulus::<$field>();
+            field_tests::run_unaryop_test_cases(
+                &modulus, WORD_BITS,
+                |x| <$field>::mul(x, x),
+                |x| x.clone() * x % &modulus)
+        }
+
+        #[test]
+        #[ignore]
+        fn arithmetic_division() -> Result<()> {
+            // This test takes ages to finish so is #[ignore]d by default.
```

Comment:
https://github.com/rust-num/num-bigint/issues/60 should make this better some day :)

Reply:
Wow, that's taking ages! I've added a comment referencing the issue.